rpc.go

```diff
@@ -120,11 +120,20 @@ func (r *rpcClient) doRPCWithRetry(ctx context.Context, address string, msg *amq
 	// this is to avoid a potential infinite loop if the returned error
 	// is always transient and Recover() doesn't fail.
 	sendCount := 0
+	var link *rpc.Link
```

Comment:
I don't think this is necessary, and in addition, it introduces a bug.  For subsequent iterations of the `for` loop, we stomp over the value of `link` without calling `Close()`.  I think a simple `defer link.Close(ctx)` on line 131 would be enough.

Reply:
That's an elegant way to do it!

With the way the code's written it'd have to be inside of the `if` though (ie: line 132)? Otherwise we're not guaranteed that `link` is not nil.

Reply:
To clarify, revert the pending change and add the `defer link.Close(ctx)` on line 131.

Reply:
Correction, we'd need the `nil` check too.

Reply:
yes that's why it's in a func. so that change is correct right?

Reply:
Ah, I see. there is still a link leak in the code, just less often.
will think about it later. thanks @jhendrixMSFT

Reply:
No, because the value of `link` is overwritten on each loop iteration, and the values used in the `defer` are evaluated when the `defer` is evaluated.  I think it should look something like this (error case omitted for brevity).
```go
var link *rpc.Link
defer func() {
    if link != nil {
        link.Close(ctx)
    }
}()
```

Reply:
the better way is to extract each iteration as a func with a defer, or to just wrap it with an anonymous func.
will send an attempt

Reply:
I'll push a correct fix tomorrow (implemented just didn't have time to self-review/push)
Your example is still leaking links when the It loops more than once, and defer in a loop is also against go guidelines.

Reply:
The defers should stack, and when the func exits they all execute in LIFO order.  At any rate, it doesn't have to be fixed like that.

Reply:
sure, but go fmt/vet stil warns about having a defer in a loop
the "recomended" way is to wrap the content of the loop in a func, and have the defer in the func.